core/controllers/base.py

```diff
@@ -282,6 +283,52 @@ def render_downloadable_file(self, values, filename, content_type):
             'attachment; filename=%s' % filename)
         self.response.write(values)
 
+    def prepare_headers(self, iframe_restriction):
+        """Prepares the headers for a HTML response.
+        Args:
+            iframe_restriction: str or None. Possible values are
+                'DENY' and 'SAMEORIGIN':
+
+                DENY: Strictly prevents the template to load in an iframe.
+                SAMEORIGIN: The template can only be displayed in a frame
+                    on the same origin as the page itself.
+        """
+        self.response.cache_control.no_cache = True
+        self.response.cache_control.must_revalidate = True
+        self.response.headers['Strict-Transport-Security'] = (
+            'max-age=31536000; includeSubDomains')
+        self.response.headers['X-Content-Type-Options'] = 'nosniff'
+        self.response.headers['X-Xss-Protection'] = '1; mode=block'
+
+        if iframe_restriction is not None:
+            if iframe_restriction in ['SAMEORIGIN', 'DENY']:
+                self.response.headers['X-Frame-Options'] = str(
+                    iframe_restriction)
+            else:
+                raise Exception(
+                    'Invalid X-Frame-Options: %s' % iframe_restriction)
+
+        self.response.expires = 'Mon, 01 Jan 1990 00:00:00 GMT'
+        self.response.pragma = 'no-cache'
+
+
+    def render_static_template(self, filepath, iframe_restriction='DENY'):
+        """Prepares a Jinja-less HTML response to be sent to the client.
+        Args:
+            filepath: str. The template filepath.
+            iframe_restriction: str or None. Possible values are
+                'DENY' and 'SAMEORIGIN':
+
+                DENY: Strictly prevents the template to load in an iframe.
+                SAMEORIGIN: The template can only be displayed in a frame
+                    on the same origin as the page itself.
+        """
+        self.prepare_headers(iframe_restriction)
+
+        path = os.path.join(
+            feconf.FRONTEND_TEMPLATES_DIR, filepath)
+        self.response.write(template.render(path, None))
```

Comment:
Just wondering ... what does template.render do? Does it have any unnecessary overhead? Is there a reference link you can point to that says "this is the preferred way to serve static files"?

Reply:
I don't think this is "the preferred way" to serve static files, but this is what was in the documentation and I expect that since it comes bundled with webapp, then it should have been optimized for the purpose.
https://webapp2.readthedocs.io/en/latest/tutorials/gettingstarted/templates.html

Reply:
template.render is similar to rendering with jinja, but it uses Django templating instead, so I'm not sure it's what you want to do here if your aim is to serve the files statically.

If you do want to serve a file statically, then typically that is done in app.yaml -- see https://cloud.google.com/appengine/docs/standard/python/config/appref

Reply:
(You might also want to try the three ways -- jinja render, django template.render, and app.yaml static rendering -- and compare their times.)

Reply:
app.yaml - 42ms
Jinja render - 147ms
template.render - 119ms

So we should go with app.yaml, but how about the headers?